File: C_case/addVec.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "func.h"

int main(int argc, char* argv[]){

    double *a, *b, *c;
    int i, nelements;

    if( argc == 2){
        nelements = atoi(argv[1]);
    }else{
        nelements = 1000;
    }

    printf(" Sum of two vectors:\n");
    printf("    c[i] = a[i] + b[i]\n");
    printf("    Test for %6d elements\n\n",nelements);

    // Allocate memory

    a = (double*) malloc(nelements * sizeof(double));
    b = (double*) malloc(nelements * sizeof(double));
    c = (double*) malloc(nelements * sizeof(double));

    assert(a);
    assert(b);
    assert(c);

    // Fill the vector
    for(i = 0; i < nelements; i++){
        a[i] = 1.;
        b[i] = 2.;
        c[i] = 0.;
    }

    // Call the sum function
    sum(nelements, a, b, c);

    // Check the sum

    check_sum(nelements, c);

    free(a);
    free(b);
    free(c);

    exit(EXIT_SUCCESS);
}

File: Fort_MKL_case/addVec.f90
program main
    use func_module
    implicit none
    integer :: nelements, i
    character(len=32) :: arg
    real*8, allocatable :: a(:), b(:), c(:)

    if (command_argument_count() == 1) then
        call get_command_argument(1, arg)
        read(unit=arg,fmt=*) nelements
    else
        nelements = 1000
    end if

    write(*,*) " Sum of two vectors:"
    write(*,*) "    c(i) = a(i) + b(i)"
    write(*,*) "    Test for", nelements, "elements"

    ! Allocate memory
    allocate(a(nelements), b(nelements), c(nelements))

    ! Fill the vectors
    a = 1.0d0
    b = 2.0d0
    c = 0.0d0

    ! Call the sum function
    call sum(nelements, a, b, c)

    ! Check the sum
    call check_sum(nelements, c)

    ! Deallocate memory
    deallocate(a, b, c)

end program main
